crates/autopilot/src/boundary/events/cow_amm_factory.rs

```diff
@@ -0,0 +1,104 @@
+use {
+    contracts::CowAmmConstantProductFactory,
+    ethcontract::{common::DeploymentInformation, dyns::DynWeb3},
+    ethrpc::current_block::RangeInclusive,
+    primitive_types::H160,
+    shared::{event_handling::EventStoring, impl_event_retrieving},
+    std::{
+        collections::{BTreeMap, HashSet},
+        sync::Arc,
+    },
+    tokio::sync::RwLock,
+};
+
+impl_event_retrieving! {
+    pub CowAmmConstantProductFactoryContract for contracts::cow_amm_constant_product_factory
+}
+
+/// CoW AMM indexer which stores events in-memory.
+#[derive(Clone)]
+pub struct Indexer {
+    /// Registry with the format: (block number, event)
+    registry: Arc<
+        RwLock<BTreeMap<u64, contracts::cow_amm_constant_product_factory::event_data::Deployed>>,
+    >,
+    first_block: u64,
+}
+
+impl Indexer {
+    pub async fn new(web3: &DynWeb3) -> Self {
+        let cow_amm_constant_product_factory = CowAmmConstantProductFactory::deployed(web3)
+            .await
+            .expect("Failed to find deployed CowAmmConstantProductFactory");
+        let first_block = match cow_amm_constant_product_factory
+            .deployment_information()
+            .expect("Failed to get deployment information")
+        {
+            DeploymentInformation::BlockNumber(block) => block,
+            _ => panic!("Expected block number"),
+        };
+        Self {
+            registry: Arc::new(RwLock::new(BTreeMap::new())),
+            first_block,
+        }
+    }
+
+    pub async fn get_cow_amm_addresses(&self) -> HashSet<H160> {
+        self.registry
+            .read()
+            .await
+            .values()
+            .map(|event| event.amm)
+            .collect()
+    }
+}
+
+#[async_trait::async_trait]
+impl EventStoring<contracts::cow_amm_constant_product_factory::Event> for Indexer {
+    async fn replace_events(
+        &mut self,
+        events: Vec<ethcontract::Event<contracts::cow_amm_constant_product_factory::Event>>,
+        range: RangeInclusive<u64>,
+    ) -> anyhow::Result<()> {
+        // Remove events in the specified range
+        let range = *range.start()..=*range.end();
```

Comment:
Why do we still need it? Can we iterate the `range: RangeInclusive<u64>` directly?

Reply:
@squadgazzz because it is not the `std` `RangeInclusive`, but a custom one introduced by the trait:

```rust
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct RangeInclusive<T: Ord> {
    start: T,
    end: T,
}
```